gcc fails on typedef of iree_atomic_intptr_t

Building IREE with gcc 8.3 or 9.3, the build fails due to 
```
/home/mbrehler/repos/iree/iree/base/atomics.h:72:9: error: ‘_Atomic’ does not name a type
   72 | typedef _Atomic __INTPTR_TYPE__ iree_atomic_intptr_t;
      |         ^
```

This results in
```
/home/mbrehler/repos/iree/iree/base/atomics.h:97:22: error: ‘iree_atomic_intptr_t’ was not declared in this scope; did you mean ‘iree_atomic_store’?
   97 | static_assert(sizeof(iree_atomic_intptr_t) == sizeof(intptr_t),
      |                      ^~~~~~~~~~~~~~~~~~~~
      |                      iree_atomic_store
```
and further the *missing* `iree_atomic_intptr_t` does effect `iree/vm/module.h` and `iree/vm/ref.h`.

Reproducible with #1268 merged.

--- a/iree/base/atomics.h
+++ b/iree/base/atomics.h
@@ -39,7 +39,7 @@ extern "C" {
 #endif
 
 #if defined(IREE_COMPILER_CLANG)
-
+// Emulate C11 atomics with builtins.
 typedef _Atomic intptr_t iree_atomic_intptr_t;
 #define IREE_ATOMIC_VAR_INIT(value) (value)
 #define iree_atomic_load(object) __c11_atomic_load(object, __ATOMIC_SEQ_CST)
@@ -51,7 +51,7 @@ typedef _Atomic intptr_t iree_atomic_intptr_t;
   __c11_atomic_fetch_sub(object, operand, __ATOMIC_SEQ_CST)
 
 #elif defined(IREE_COMPILER_MSVC)
-
+// Emulate C11 atomics with Interlocked win32 APIs.
 // NOTE: currently assumes sizeof(intptr_t) == 8.
 typedef struct {
   intptr_t __val;
@@ -68,19 +68,27 @@ typedef struct {
   InterlockedExchangeAdd64((volatile LONGLONG*)object, -(operand))
 
 #elif defined(IREE_COMPILER_GCC)
-
-typedef _Atomic __INTPTR_TYPE__ iree_atomic_intptr_t;
+// Emulate atomics for GCC in a way that is compatible for inclusion in
+// both C and C++ modes.
+#ifdef __cplusplus
+// Equiv to C++ auto keyword in C++ mode.
+#define __iree_auto_type auto
+#else
+// Only defined in C mode.
+#define __iree_auto_type __auto_type
+#endif
+typedef __INTPTR_TYPE__ iree_atomic_intptr_t;
 #define IREE_ATOMIC_VAR_INIT(value) (value)
 #define iree_atomic_load(object)                                              \
   __atomic_load_ptr(object, __ATOMIC_SEQ_CST) __extension__({                 \
-    __auto_type __atomic_load_ptr = (object);                                 \
+    __iree_auto_type __atomic_load_ptr = (object);                            \
     __typeof__(*__atomic_load_ptr) __atomic_load_tmp;                         \
     __atomic_load(__atomic_load_ptr, &__atomic_load_tmp, (__ATOMIC_SEQ_CST)); \
     __atomic_load_tmp;                                                        \
   })
 #define iree_atomic_store(object, desired)                          \
   __extension__({                                                   \
-    __auto_type __atomic_store_ptr = (object);                      \
+    __iree_auto_type __atomic_store_ptr = (object);                 \
     __typeof__(*__atomic_store_ptr) __atomic_store_tmp = (desired); \
     __atomic_store(__atomic_store_ptr, &__atomic_store_tmp,         \
                    (__ATOMIC_SEQ_CST));                             \

--- a/iree/modules/tensorlist/native_module.cc
+++ b/iree/modules/tensorlist/native_module.cc
@@ -322,12 +322,14 @@ class TensorList final : public RefObject<TensorList> {
 static iree_vm_ref_type_descriptor_t iree_tensorlist_descriptor = {0};
 
 // Register our type with the vm::ref<T> static machinery.
+namespace vm {
 template <>
-struct ::iree::vm::ref_type_descriptor<TensorList> {
+struct ref_type_descriptor<TensorList> {
   static const iree_vm_ref_type_descriptor_t* get() {
     return &iree_tensorlist_descriptor;
   }
 };
+}  // namespace vm
 
 extern "C" iree_status_t iree_tensorlist_module_register_types() {
   static bool has_registered = false;

